PRODDOCS-521: Unions+intersections (#8898)
fishtown-analytics/docs.getdbt.com: website/docs/reference/node-selection/set-operators.md

---
title: "Set operators"
description: "Union and intersection set operators for dbt --select and --exclude."
---

> Unions and intersections combine <Term id="selector-expression">selector expressions</Term> when you list multiple arguments for `--select` or `--exclude`. A selector expression defines a subset of nodes in your project (models, tests, seeds, and other resource types) so you do not have to run the entire DAG. For the full selection language and for named selectors in `selectors.yml`, refer to [Syntax overview](/reference/node-selection/syntax) and [YAML selectors](/reference/node-selection/yaml-selectors).

Reviewer: Hey Nat, good idea to add a hoversnippet here! However, this feels a bit repetitive as written. The paragraph explains what a selector expression is ("defines a subset of nodes…") and its purpose, and the hoversnippet somewhat repeats that definition. 

It might help to make the hoversnippet shorter and more complementary, since longer definitions can be harder to scan when you hover. For example, we could simplify to:
> An expression used with `--select` and `--exclude` to include or exclude specific nodes in your dbt project.

## Unions

> A union merges the node sets from several selector expressions: every resource that matches any of those expressions is kept. In other words, union behaves like OR across the arguments you pass to `--select` or `--exclude`.

Reviewer: Suggesting to change to semi-colon to be consistent with line 24, and to add emphasis to `any`
```suggestion
A union merges the node sets from several selector expressions; every resource that matches _any_ of those expressions is kept. In other words, union behaves like OR across the arguments you pass to `--select` or `--exclude`.
```

> A union merges the node sets from several selector expressions; every resource that matches _any_ of those expressions is kept. In other words, union behaves like OR across the arguments you pass to `--select` or `--exclude`.

Reviewer: ```suggestion
A union merges the node sets from several selector expressions; every resource that matches _any_ of those expressions is selected. In other words, union behaves like OR across the arguments you pass to `--select` or `--exclude`.
```

> With unions, you can pass multiple arguments separated by spaces (space-delimited). dbt resolves each argument using the normal selection rules ([selection methods](/reference/node-selection/methods), [graph operators](/reference/node-selection/graph-operators), and so on), then combines the results. If more than one argument matches the same node, that node still appears only once in the final selection (there are no duplicates).

Reviewer: "and so on" is a bit vague
```suggestion
With unions, you can pass multiple arguments separated by spaces (space-delimited). dbt resolves each argument using the normal selection rules ([selection methods](/reference/node-selection/methods), [graph operators](/reference/node-selection/graph-operators), and other selection syntax), then combines the results. If more than one argument matches the same node, that node still appears only once in the final selection (there are no duplicates).
```

> With unions, you can pass multiple arguments separated by spaces (space-delimited). dbt resolves each argument using the normal selection rules ([selection methods](/reference/node-selection/methods), [graph operators](/reference/node-selection/graph-operators), and other selection syntax), then combines the results. If more than one argument matches the same node, that node still appears only once in the final selection (there are no duplicates).

Reviewer: ```suggestion
With unions, you can pass multiple arguments separated by spaces (space-delimited). dbt resolves each argument using the normal selection rules ([selection methods](/reference/node-selection/methods), [graph operators](/reference/node-selection/graph-operators), and other selection syntax), then combines the results. If more than one argument matches the same node, that node appears only once in the final selection.
```

> For example, the following command unions two selector expressions. Each uses `+` to include a model and its ancestors, and the space between them merges both sets:

Reviewer: Using "unions" as a verb reads a bit awkward -- would it make sense to change to:

```suggestion
For example, the following command combines two selector expressions. Each uses `+` to include a model and its ancestors, and the space between them merges both sets:
```

```bash
dbt run --select "+snowplow_sessions +fct_orders"
```

> That differs from an [intersection](#intersections), where comma-separated arguments with no spaces between them require a resource to satisfy _all_ criteria at once.

Reviewer: ```suggestion
This behavior differs from an [intersection](#intersections), where comma-separated arguments with no spaces between them require a resource to satisfy _all_ criteria at once.
```

## Intersections

> An intersection keeps only the nodes that appear in every selector result; a resource must match every comma-separated argument to stay selected. If it does not match any one of those arguments, it is excluded from the final set.

Reviewer: ```suggestion
An intersection keeps only the nodes that match every selector expression; a resource must satisfy _all_ comma-separated arguments to remain in the final set.
```

> Use commas with no spaces between them to request an intersection when you pass multiple arguments to `--select` or `--exclude`. Spaces between arguments still mean a [union](#unions). dbt resolves each argument using the normal selection rules ([selection methods](/reference/node-selection/methods), [graph operators](/reference/node-selection/graph-operators), and so on), then keeps only resources that satisfy _all_ of them. The order of comma-separated arguments does not change the final set.

Reviewer: ```suggestion
Use commas with no spaces between arguments to define an intersection when you pass multiple arguments to `--select` or `--exclude`. Spaces between arguments still mean a [union](#unions). dbt resolves each argument using the normal selection rules ([selection methods](/reference/node-selection/methods), [graph operators](/reference/node-selection/graph-operators), and other selection syntax), then keeps only resources that satisfy all of them. The order of comma-separated arguments does not change the final set.
```

Select shared upstream nodes (common ancestors of `snowplow_sessions` and `fct_orders`):

```bash
dbt run --select "+snowplow_sessions,+fct_orders"
```

> 

Reviewer: I suggest we add an intro here that these are samples of intersections

Select shared downstream nodes (common descendants of `stg_invoices` and `stg_accounts`):

```bash
dbt run --select "stg_invoices+,stg_accounts+"
```

Select models that are under the `marts/finance` path and tagged `nightly`:

```bash
dbt run --select "marts.finance,tag:nightly"
```

## Combining unions and intersections

> You can combine unions and intersections in a single `--select` or `--exclude` value. dbt evaluates each space-delimited argument independently; within each argument, commas with no spaces create an intersection. A space between two arguments combines their results into a union, letting you apply different filtering logic to different subsets of your project in a single command.

Reviewer: Suggest to lead with what they can do when they combine unions and intersections, let me know what you think!
```suggestion
You can combine unions and intersections in a single `--select` or `--exclude` value to build complex selections in one command. dbt evaluates each space-delimited argument independently &mdash; commas with no spaces within an argument define an intersection, and a space between arguments combines their results as a union. Combining both operators lets you apply different filtering logic to different subsets of your project at once.
```

> You can combine unions and intersections in a single `--select` or `--exclude` value. dbt evaluates each space-delimited argument on its own; commas with no spaces still create intersections within that argument. Joining two expressions with a space unions their results, so you can intersect to narrow one slice of the graph while still including another slice in the same run. That reduces how much of the project runs compared to a broader selection, without dropping nodes you still need.

Reviewer: ```suggestion
You can combine unions and intersections in a single `--select` or `--exclude` value. dbt evaluates each space-delimited argument independently; within each argument, commas with no spaces create an intersection. A space between two arguments combines their results into a union, letting you apply different filtering logic to different subsets of your project in a single command.
```

> For example, the following command combines two intersections: shared upstream nodes for `snowplow_sessions` and `fct_orders`, and models under `marts/finance` tagged `nightly`:

Reviewer: Maybe we can explain this example a bit more. Something like:

For example, the following command contains two space-separated arguments, each defining an intersection:

- `+snowplow_sessions,+fct_orders`: selects nodes that are upstream of both `snowplow_sessions` and `fct_orders`
- `marts.finance,tag:nightly`: selects models that are under the `marts/finance` path and tagged `nightly`

The space between them creates a union and combines the two results into a single selection:

dbt run --select "+snowplow_sessions,+fct_orders marts.finance,tag:nightly"

> For example, the following command unions two intersections: shared upstream nodes for `snowplow_sessions` and `fct_orders`, and models under `marts/finance` tagged `nightly`.

Reviewer: ```suggestion
For example, the following command combines two intersections: shared upstream nodes for `snowplow_sessions` and `fct_orders`, and models under `marts/finance` tagged `nightly`:
```
